apps/dotcom/src/hooks/useMultiplayerAssets.ts

```diff
@@ -58,7 +43,68 @@ export function useMultiplayerAssets(assetUploaderUrl: string) {
 					isAnimated,
 				},
 				meta: {},
-			})
+			} satisfies TLVideoAsset) as TLVideoAsset
+
+			return asset
+		},
+		[assetUploaderUrl]
+	)
+}
+
+export function useMultiplayerImageAsset(assetUploaderUrl: string) {
+	return useCallback(
+		async (files: File[]): Promise<TLImageAsset> => {
+			const id = uniqueId()
+
+			const UPLOAD_URL = `${assetUploaderUrl}/uploads`
+
+			const urls: string[] = []
+
+			let size: { w: number; h: number } | undefined
+			let isAnimated = false
+			let assetId: TLAssetId | undefined
+			let name: string | undefined
+			let mimeType: string | undefined
+
+			// let's assume the scale will be 1, .5, .25, with the first being the largest
+
+			for (let i = 0; i < files.length; i++) {
+				const file = files[i]
+				if (!DEFAULT_ACCEPTED_IMG_TYPE.includes(file.type)) throw Error('File is not an image')
+
+				const objectName = `${id}-${file.name}-${i}`.replaceAll(/[^a-zA-Z0-9.]/g, '-')
+				const url = `${UPLOAD_URL}/${objectName}`
+
+				if (i === 0) {
+					name = file.name
+					assetId = AssetRecordType.createId(getHashForString(url))
+					if (file.type === 'image/gif') isAnimated = true
+					size = await MediaHelpers.getImageSize(file)
+				}
+
+				urls.push(url)
+				await fetch(url, {
+					method: 'POST',
+					body: file,
+				})
```

Comment:
can we tweak this to perform uploads in parallel?

Reply:
if we did this in parallel it feels like that might be potentially harder on bandwidth, perhaps for certain users?

Reply:
fair! tbh for this multiplayer one we could always look at using https://developers.cloudflare.com/images/transform-images/ if we wanted to really optimize this - just upload the original, let the server handle resizing on-demand (plus optimizing for smaller file size in a way that the browser won't out of the box)

Reply:
Hahah, Steve and I were indeed talking about this same thing this morning (albeit with Imgix, not Cloudflare's offering). I'll pull you aside and we can chat more offline on your thoughts on this.